### PR TITLE
bpo-43892: Make match patterns explicit in AST

### PR DESCRIPTION
**Pull request by ncoghlan**
* Separates "pattern" and "expr" nodes in AST
* Adds a new "Pattern matching" section in the ast docs,
  separate from the "Control flow" section
* AST node names are inspired by the AST proposed
  in PEP 642, but adjusted to better match the semantics
  of PEP 634 and to be more consistent with the existing AST
* Grammar definition has been updated to emit the
  newly defined AST nodes
* Parser now emits SyntaxError for ill-formed complex literals
  (so the AST optimiser no longer has to do it)
* Code generator consumes the new nodes
* AST validator consumes the new nodes
* AST unparser consumes new nodes
* test_unparse always checks the AST roundtrip for test_patma.py
* Add some clarifying comments to the unparsing code

https://bugs.python.org/issue43892

### CODE REVIEW

**Comment by ncoghlan** [comment]
* Merged in master, adjusting for the recursion tracking in the AST validator
* Updated the AST documentation (the MatchAs and MatchOr doctests failed, highlighting that the new nodes also needed documentation)

**Comment by ncoghlan** [comment]
I think I worked out what must have been going on with the `0+0` test case: I must have been testing the `0+0` rejection at the command line with `compile(ast.parse(...), ...)` rather than actually running `test_patma` repeatedly.

`test_patma` itself was still failing because the PR relies on the AST validator to reject `0+0`, but the compiler hasn't historically validated the AST emitted by its own parser - it has assumed it is always going to be valid. That assumption isn't true for pattern matching, as the parser can't distinguish between real and imaginary numbers, so it can't detect that "0+0" isn't a valid complex literal in a pattern. I've modified compilation from strings to ensure the emitted AST is valid, so `test_patma` now passes as expected.

The original pattern matching implementation used the AST optimiser pass to pick up that case, but this really isn't the optimiser's job to prevent.

**Comment by ncoghlan** [comment]
Unconditionally validating the parser output showed that the validator wasn't actually handling FunctionType nodes (and this was one of the compiler switch statements that I *hadn't* migrated over to emitting a compile-time warning for unhandled enum values).

**Comment by isidentical** [comment]
> Unconditionally validating the parser output showed that the validator wasn't actually handling FunctionType nodes

I think that was intentional since the only place (`compile()`) `PyAST_Validate` called was rejecting `FunctionType`'s beforehand.

**Comment by ncoghlan** [comment]
Following through on @isidentical's comment, the way FunctionType nodes in the AST have historically been handled:
* they would fail validation if you tried to compile them to code
* compiling from strings didn't attempt to validate them because it never attempted to validate anything
* compiling from AST with the compile() builtin didn't attempt to validate them if the PyCF_ONLY_AST flag was set (as that also doesn't attempt to validate anything), but *would* attempt to validate them if full code generation was requested and a compilation mode other than "func_type" was set (and the validation would then fail).

With the PR, they'll pass the AST validation step if they're well-formed FunctionType nodes, but actual compilation will still fail. It will just fail on the symtable analysis pass, as that also rejects them.

File: Doc/library/ast.rst
Comment (ncoghlan) [inline Doc/library/ast.rst]: This section of the docs could use some copy-editing (e.g. I don't think this sentence is needed any more now that the dedicated nodes exist), but I'd prefer to leave that to a later PR rather than trying to clean it up alongside the functional change.

File: Python/compile.c
Comment (ncoghlan) [inline Python/compile.c:5868]: I'm not sure this is still right given @pablogsal's recent syntax error reporting changes, so I suspect this code may need an error reporting cleanup pass post-beta.

Comment (pablogsal) [reply Python/compile.c:5868]: You just need to take the other new fields into account. If you rebase against the latest master you will see examples in this file. Otherwise I can do it after you merge the PR :)

File: Lib/ast.py
Comment (isidentical) [inline Lib/ast.py:803]: `visit()` as in `_Unparser` is just the top level entry point, methods within this class should use `traverse()` to rely on actual `NodeVisitor.generic_visit`. Though I think this might be something we should change in the future if we ever decide to publicize the API (I get the confusion).

File: Doc/library/ast.rst
Comment (brandtbucher) [inline Doc/library/ast.rst]:
```suggestion
   If ``name`` is not ``None``, a list containing the remaining sequence
```

File: Doc/library/ast.rst
Comment (brandtbucher) [inline Doc/library/ast.rst]:
```suggestion
   match pattern that will be used to match the subject against.
```

File: Doc/library/ast.rst
Comment (brandtbucher) [inline Doc/library/ast.rst]:
```suggestion
        ...     case {1: _, 2: _}:
```

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
        if node.pattern is not None:
```

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
                lambda: self.write(", "), write_key_pattern_pair, zip(keys, node.patterns, strict=True)
```

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
                    lambda: self.write(", "), write_attr_pattern, zip(attrs, node.kwd_patterns, strict=True)
```

File: Python/symtable.c
Comment (brandtbucher) [inline Python/symtable.c]:
```suggestion
        VISIT_SEQ(st, expr, p->v.MatchMapping.keys);
        VISIT_SEQ(st, pattern, p->v.MatchMapping.patterns);
```

File: Python/symtable.c
Comment (brandtbucher) [inline Python/symtable.c]:
```suggestion
        VISIT_SEQ(st, pattern, p->v.MatchClass.patterns);
        VISIT_SEQ(st, pattern, p->v.MatchClass.kwd_patterns);
```

**Review by brandtbucher — COMMENTED** [review]
Thanks for taking the time to do all of this @ncoghlan!

I wasn't able to finish reviewing this tonight, but I've left some comments on the files I did get to (I can also just apply them before merging if you can't get to them).

I still need to finish reviewing:

- ` Grammar/python.gram`
- ` Python/ast.c`
- ` Python/ast_opt.c`
- ` Python/compile.c`

I should be able to get to them tomorrow.

Also (something I thought about during review), how do you feel about ditching `MatchAlways` and making the `name` member of `MatchAs` optional? That way `_` can be represented as just `MatchAs(pattern=None, name=None)`. I think that should simplify checks for irrefutable patterns.

If you think that's an improvement (I think I do) we can do it in a follow-up PR.

File: Doc/library/ast.rst
Comment (ncoghlan) [reply Doc/library/ast.rst]: I replaced this with a sentence about the pattern AST using different nodes for the same syntax.

File: Doc/library/ast.rst
Comment (ncoghlan) [reply Doc/library/ast.rst]: Huh, I had managed to miss that in both the PEPs and the implementation. I guess it makes it easier to accumulate the extra elements and avoids having to copy them into a tuple at the end.

File: Lib/ast.py
Comment (ncoghlan) [reply Lib/ast.py:803]: Right, but the issue is that `NodeVisitor.generic_visit` itself calls `self.visit()` rather than `self.traverse()`, so if `_Unparser` doesn't handle every node you lose all of your previous output. As it turned out, the generic visitor was wrong for `MatchSingleton` anyway, and `MatchValue` is a one-liner, so not relying on it didn't cost me much.

File: Lib/ast.py
Comment (ncoghlan) [reply Lib/ast.py]: Reworking this method as part of replacing MatchAlways() with MatchAs()

File: Python/compile.c
Comment (ncoghlan) [reply Python/compile.c:5868]: I'm going to take you up on the offer of a post-merge cleanup :)

**Comment by ncoghlan** [comment]
@brandtbucher I tried out the MatchAlways -> MatchAs(None, None) change and it was straightforward to document and involved deleting more code than it added, so I went ahead and made that update after incorporating your other suggestions (at least, the ones that didn't touch the same parts of the code or docs as the other change).

**Comment by ncoghlan** [comment]
After accidentally checking in the doctest output artifacts twice while working on this PR, I've added a Doc/.gitignore file so that won't happen in the future.

**Comment by ncoghlan** [comment]
Windows failure looks like https://bugs.python.org/issue41682

I haven't triggered a re-run, as CI will re-run anyway if @brandtbucher has any further comments to incorporate.

File: Misc/NEWS.d/next/Core and Builtins/2021-04-25-08-35-11.bpo-43892.hr5Ke2.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2021-04-25-08-35-11.bpo-43892.hr5Ke2.rst]:
```suggestion
Match patterns now use new dedicated AST nodes (``MatchValue``,
```

File: Doc/.gitignore
Comment (brandtbucher) [inline Doc/.gitignore]: I'm going to remove the new `.gitignore` from this PR, since it looks like it might have gotten messed up by your most recent commit.

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
            self.interleave(
                lambda: self.write(", "), self.traverse, node.patterns
            )
```

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
                lambda: self.write(", "),
                write_key_pattern_pair,
                zip(keys, node.patterns, strict=True),
```

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]:
```suggestion
                    lambda: self.write(", "),
                    write_attr_pattern,
                    zip(attrs, node.kwd_patterns, strict=True),
```

File: Python/ast.c
Comment (brandtbucher) [inline Python/ast.c]:
```suggestion
            if (p->v.MatchAs.pattern == NULL) {
                ret = 1;
            }
            else if (p->v.MatchAs.name == NULL) {
                PyErr_SetString(PyExc_ValueError,
                                "MatchAs must specify a target name if a pattern is given");
                return 0;
            }
            else {
                ret = validate_pattern(state, p->v.MatchAs.pattern);
```

File: Python/compile.c
Comment (brandtbucher) [inline Python/compile.c]:
```suggestion
        assert((i == star) == (pattern->kind == MatchStar_kind));
```

File: Python/compile.c
Comment (brandtbucher) [inline Python/compile.c]:
```suggestion
    Py_ssize_t nargs = asdl_seq_LEN(patterns);
    Py_ssize_t nattrs = asdl_seq_LEN(kwd_attrs);
    Py_ssize_t nkwd_patterns = asdl_seq_LEN(kwd_patterns);
```

**Review by brandtbucher — APPROVED** [review]
Awesome, done reviewing. Just a few style nits and cleanup from refactoring:

**Comment by brandtbucher** [comment]
@pablogsal, do you want to see buildbots on this before merging?

**Comment by pablogsal** [comment]
> @pablogsal, do you want to see buildbots on this before merging?

Either that or run manually a bunch of related tests yourself with -R : and paste the output here :)

**Comment by brandtbucher** [comment]
This good?

```
$ ./python -m test -R : test_ast test_symtable test_peg_generator test_patma test_grammar test_unparse test_compile test_syntax
0:00:00 load avg: 2.04 Run tests sequentially
0:00:00 load avg: 2.04 [1/8] test_ast
beginning 9 repetitions
123456789
.........
0:00:21 load avg: 1.82 [2/8] test_symtable
beginning 9 repetitions
123456789
.........
0:00:21 load avg: 1.82 [3/8] test_peg_generator
beginning 9 repetitions
123456789
/home/bucher/src/cpython/Lib/test/test_peg_generator/test_c_parser.py:4: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  from distutils.tests.support import TempdirManager
/home/bucher/src/cpython/Lib/test/support/__init__.py:1671: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
  from distutils import ccompiler, sysconfig, spawn, errors
.........
0:08:10 load avg: 1.41 [4/8] test_patma -- test_peg_generator passed in 7 min 48 sec
beginning 9 repetitions
123456789
.........
0:08:10 load avg: 1.41 [5/8] test_grammar
beginning 9 repetitions
123456789
.........
0:08:10 load avg: 1.41 [6/8] test_unparse
beginning 9 repetitions
123456789
.........
0:08:36 load avg: 1.40 [7/8] test_compile
beginning 9 repetitions
123456789
.........
0:09:14 load avg: 1.27 [8/8] test_syntax -- test_compile passed in 38.1 sec
beginning 9 repetitions
123456789
.........

== Tests result: SUCCESS ==

All 8 tests OK.

Total duration: 9 min 15 sec
Tests result: SUCCESS

```

**Comment by pablogsal** [comment]
> enerator test_patma test_grammar test_unparse test_compile test_syntax
> 0:00:00 load avg: 2.04 Run tests sequentially
> 0:00:00 load avg: 2.04 [1/8] test_ast

Yup yup, go ahead! :rocket:

File: Python/pythonrun.c
Comment (pablogsal) [inline Python/pythonrun.c]: Wait, why is this change here? THis is going to slow down quite a lot `Py_CompileStringObject` code path. Notice that the parser already calls this on input only on debug mode precisely because of performance concerns. 

The design principle is that the ast that comes from the parser must be always correct and I think we should certainly preserve that.

File: Grammar/python.gram
Comment (pablogsal) [inline Grammar/python.gram:376]: You can probably change this to just `','.keyword_pattern+`. Same for similar stuff around this like `| items=','.key_value_pattern+ { items }`. But we can clean this afterwards, for sure.

File: Python/pythonrun.c
Comment (brandtbucher) [reply Python/pythonrun.c]: Let's remove it for now, then. I understood it as being motivated by the points in [this comment](https://github.com/python/cpython/pull/25585#issuecomment-826926615).

Comment (pablogsal) [reply Python/pythonrun.c]: Regading this comment there is an important fact to highlight:

> compiling from strings didn't attempt to validate them because it never attempted to validate anything

The actual reason is because by design the parser code *must* return a valid AST (there is a debug-mode validation step to ensure that this contract is true).

Comment (brandtbucher) [reply Python/pythonrun.c]: Hm, looks like the new code might have been relying on that pass to give us a SyntaxError in `test_patma_240`. I'll try to figure out what's going on.

**Comment by brandtbucher** [comment]
Okay, I disabled some overly-aggressive optimizations for value patterns and mapping patterns in the AST optimizer and moved validation of numeric literals back into the compiler. Reran the refleak tests locally and everything looks good.

**Comment by brandtbucher** [comment]
@ncoghlan, it's almost time for me to go to sleep and for you to wake up. I'm happy with this now, so go ahead and merge if you agree!

File: Python/pythonrun.c
Comment (ncoghlan) [reply Python/pythonrun.c]: The parser can't verify that a complex literal is actually a complex literal because the tokeniser just emits `NUMBER` for both real and imaginary numbers. As a result, `0+0` makes it through the parsing rules for pattern matching.

The compiler can't reject the syntax because it never sees it - the value has turned into the legal constant `0` by the time the pattern gets to it.

The original pattern matching implementation worked around that limitation by putting a validation step in the AST constant folding pass: instead of using the standard constant folding for expressions, it reimplemented the subset required to fold the literals that pattern matching supported and emitted a syntax error for binary operations that weren't complex numbers.

This PR switches over to using the standard expression folding functions in the optimisation pass, so forcing validation was needed to keep "0+0" illegal in value patterns and mapping pattern keys.

I'll try replace it with a more targeted fix that adds a helper function to pegen that ensures the AST coming out of the parser is valid rather than relying on the AST validator or the AST constant folding to reject it.

Comment (pablogsal) [reply Python/pythonrun.c]: > The compiler can't reject the syntax because it never sees it - the value has turned into the legal constant 0 by the time the pattern gets to it.

We probably shouldn't fold the AST for patterns in that case so the compiler can handle it. Or I am missing something?

Comment (ncoghlan) [reply Python/pythonrun.c]: I just pushed a change that restores the invariant of the parser always producing a valid AST without having to weaken the checks on the AST nodes for value patterns and mapping patterns keys.

The trick is a new `_PyAST_EnsureImaginary` helper function that accepts and returns `expr_ty`.

It looks like I didn't get the reversion/merge with Brandt's changes quite right though, so I'll post an example after cleaning that up.

Comment (ncoghlan) [reply Python/pythonrun.c]: I had reverted the validator and compiler changes, but the AST optimiser changes needed to be reverted as well.
This is how compilation from a string behaves with the parser enhanced to ensure that complex literals actually are complex numbers:
```
>>> match x:
...     case 0+0:
SyntaxError: Imaginary number required in complex literal
```

So with that everything is handling the responsibilities it should be handling:
* parser rejects "0+0" when compiling from a string rather than producing an invalid AST
* validator rejects "0+0" when compiling from a prebuilt AST tree
* optimiser and compiler have enough checks to protect themselves from segfaults, but are otherwise pretty tolerant

**Comment by ncoghlan** [comment]
@brandtbucher The parser potentially producing an illegal AST tree never sat well with me, and it finally occurred to me today how to fix it: filter the required-to-be-imaginary parse nodes through an AST helper function that raises a syntax error if they're not complex numbers. Between the tokenizer ensuring that the nodes are literals, and the helper function ensuring that they're complex number, we get a combined assurance that the value is an imaginary number.

So I reworked things to frontload the validation again, while keeping your other changes.

File: Python/pythonrun.c
Comment (ncoghlan) [reply Python/pythonrun.c]: After the first iteration, I realised the imaginary number checking would work better as a pegen helper rather than as an AST helper:
```
>>> match x:
...     case 0+0:
  File "<stdin>", line 2
    case 0+0:
           ^
SyntaxError: Imaginary number required in complex literal
```

File: Grammar/python.gram
Comment (ncoghlan) [reply Grammar/python.gram:376]: Yeah, this and PEP 642 were my first real ventures into working with the new parser. It's really nice, but it isn't always obvious when it will need help with typecasting and when it will be able to figure it on its own (e.g. I think this node used to be a more specific type than adsl_seq, so it would have needed the typecast)

**Comment by ncoghlan** [comment]
Updated refleak hunting results (still clean, although it appears my laptop is slightly slower than Brandt's machine):

```
$ ./python -m test -R : test_ast test_symtable test_peg_generator test_patma test_grammar test_unparse test_compile test_syntax
0:00:00 load avg: 0.68 Run tests sequentially
0:00:00 load avg: 0.68 [1/8] test_ast
beginning 9 repetitions
123456789
.........
0:00:20 load avg: 0.79 [2/8] test_symtable
beginning 9 repetitions
123456789
.........
0:00:20 load avg: 0.79 [3/8] test_peg_generator
beginning 9 repetitions
123456789
/home/ncoghlan/devel/cpython/Lib/test/test_peg_generator/test_c_parser.py:4: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  from distutils.tests.support import TempdirManager
/home/ncoghlan/devel/cpython/Lib/test/support/__init__.py:1671: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
  from distutils import ccompiler, sysconfig, spawn, errors
.........
0:10:38 load avg: 1.39 [4/8] test_patma -- test_peg_generator passed in 10 min 17 sec
beginning 9 repetitions
123456789
.........
0:10:38 load avg: 1.39 [5/8] test_grammar
beginning 9 repetitions
123456789
.........
0:10:39 load avg: 1.39 [6/8] test_unparse
beginning 9 repetitions
123456789
.........
0:10:58 load avg: 1.28 [7/8] test_compile
beginning 9 repetitions
123456789
.........
0:11:31 load avg: 1.15 [8/8] test_syntax -- test_compile passed in 33.4 sec
beginning 9 repetitions
123456789
.........

== Tests result: SUCCESS ==

All 8 tests OK.

Total duration: 11 min 32 sec
Tests result: SUCCESS
```

File: Grammar/python.gram
Comment (ncoghlan) [inline Grammar/python.gram:287]: @brandtbucher @pablogsal This is the new helper that ensures the parser won't accept "0+0" and similar constructs as "complex literals".

File: Grammar/python.gram
Comment (pablogsal) [reply Grammar/python.gram:376]: > Yeah, this and PEP 642 were my first real ventures into working with the new parser. It's really nice, but it isn't always obvious when it will need help with typecasting and when it will be able to figure it on its own (e.g. I think this node used to be a more specific type than adsl_seq, so it would have needed the typecast)

I'm preparing a new documentation for working with the parser :)

How has been your experience so far? Any feedback (good and bad) is good feedback ;)

**Comment by brandtbucher** [comment]
Hm, this still allows some illegal value expressions:

```
>>> match 1j:
...    case 1j+0j: print(":(")
... 
:(
```

(We should probably add a test for this.)

I personally still prefer checking these in the compiler (as I left the branch last night), since it keeps us from having to put the same logic for rejecting illegal value patterns in two places (the parser *and* the validator). Note that there are already lots of illegal trees that make it as far as the compiler: `yield`/`async`/`await`/`return` outside of functions, `break`/`continue` outside of loops, illegal `*` expressions, etc.

I'm almost ready to drop complex literal value patterns entirely and wait until somebody asks for them. And then I can just tell them to use a dotted name instead. 😉

**Comment by pablogsal** [comment]
> Hm, this still allows some illegal value expressions:
> 
> ```py
> >>> match 1j:
> ...     case 1j+0j:
> ...         print(":(")
> ... 
> :(
> ```
> 
> (We should probably add a test for this.)
> 
> I personally still prefer checking these in the compiler (as I left the branch last night), since it keeps us from having to put the same logic for rejecting illegal value patterns in two places (the parser *and* the validator).
> 
> Note that there are already lots of illegal trees that make it as far as the compiler: `yield`/`async`/`await`/`return` outside of functions, `break`/`continue` outside of loops, illegal `*` expressions, etc. I see the job of the validator as being "make sure this tree can't crash the compiler", not "make sure this tree is valid Python".
> 


The key is that the AST check is to check manually crafted AST objects. Otherwise the compiler and the parser are obviously coupled, so the compiler knows what needs to check and what not, which is how is normally done and keeps it efficient.

The slightly sad thing that I want to improve is more Syntax errors to the grammar so the grammar represents Python and not "almost Python".

> I'm almost ready to drop complex literal value patterns entirely and wait until somebody asks for them. And then I can just tell them to use a dotted name instead. 😉 

You have until this Sunday to decide 😉

**Comment by brandtbucher** [comment]
Looping in @gvanrossum. Are complex literal value patterns worth the complexity right now? I think we just added them because we could.

If so, then we can throw out `BinOp` values entirely, and only accept:
- strings
- byte-strings
- floats
- negated floats
- ints
- negated ints
- imaginary numbers
- negated imaginary numbers

So, continue to allow (`4.2j`, and `-4.2j`, but not `-4.2+4.2j`).

**Comment by gvanrossum** [comment]
I find it important that we can spell all values for these types. So I think we should keep it. (But I do *not* think that checking for all invalid cases needs to be implemented before beta 1.)

**Comment by brandtbucher** [comment]
Okay, that makes sense.

Unfortunately, I'm not sure how to regenerate the parser on somebody else's PR. Since there's already a lot going on here, I am going to just land this as-is and open a small follow-up PR for the cleanup.

**Comment by ncoghlan** [comment]
That approach makes sense to me, too. Thanks for the reviews @brandtbucher, @pablogsal!

**Pull request merged**
